File: 42-trapping-rain-water/42-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int n = height.length, trappedWater = 0;
        if(n <= 2) return 0;
        
        int left = 1, right = n - 2, maxFromLeft = height[0], maxFromRight = height[n - 1];
        
        while(left <= right) {
            if(maxFromLeft < maxFromRight) {
                if(height[left] > maxFromLeft)
                    maxFromLeft = height[left];
                else
                    trappedWater += maxFromLeft - height[left];
                ++left;
            } else {
                if(height[right] > maxFromRight)
                    maxFromRight = height[right];
                else 
                    trappedWater += maxFromRight - height[right];
                --right;
            }
        }
        
        return trappedWater;
    }
}

File: 42-trapping-rain-water/42-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size(), trappedWater = 0;
        if(n == 1) return 0;
        
        vector<int> maxFromLeft(n), maxFromRight(n);
        maxFromLeft[0] = height[0], maxFromRight[n - 1] = height[n - 1];
        
        for(int i = 1; i < n; ++i) 
            maxFromLeft[i] = max(maxFromLeft[i - 1], height[i]);
        for(int i = n - 2; i >= 0; --i)
            maxFromRight[i] = max(maxFromRight[i + 1], height[i]);
        
        for(int i = 0; i < height.size(); ++i)
            trappedWater += min(maxFromLeft[i], maxFromRight[i]) - height[i];
        
        return trappedWater;
    }
};